File: mat/1/src/path.zig
const std = @import("std");
const types = @import("types.zig");

const Vertex = types.Vertex;

pub fn dijkstra(allocator: std.mem.Allocator, adj_matr: [][]f64, start: usize) ![]f64 {
    const n = adj_matr.len;

    var visited: []bool = try allocator.alloc(bool, n);
    defer allocator.free(visited);
    std.mem.set(bool, visited, false);

    var marks: []f64 = try allocator.alloc(f64, n);
    std.mem.set(f64, marks, std.math.inf(f64));
    marks[start] = 0;

    var queue = std.PriorityQueue(Vertex, []f64, Vertex.order).init(allocator, marks);
    try queue.add(.{ .index = start });

    while (queue.count() != 0) {
        const cur = queue.remove();

        if (visited[cur.index]) {
            continue;
        }

        visited[cur.index] = true;

        for (adj_matr[cur.index]) |dist, index| {
            if (dist == 0) {
                continue;
            }

            const new_dist = marks[cur.index] + dist;
            if (new_dist < marks[index]) {
                marks[index] = new_dist;
                try queue.add(.{ .index = index });
            }
        }
    }

    return marks;
}

pub fn fordBellman(allocator: std.mem.Allocator, adj_matr: [][]f64, start: usize) !struct {
    dist: []f64,
    found_negative: bool,
} {
    const n = adj_matr.len;

    var dist: []f64 = try allocator.alloc(f64, n);
    std.mem.set(f64, dist, std.math.inf(f64));
    dist[start] = 0;

    var i: usize = 0;
    while (i < n - 1) : (i += 1) {
        for (adj_matr) |edges, ui| {
            for (edges) |weight, vi| {
                if (weight == 0) {
                    continue;
                }

                const new_dist = dist[ui] + weight;
                if (dist[vi] > new_dist) {
                    dist[vi] = new_dist;
                }
            }
        }
    }

    for (adj_matr) |edges, ui| {
        for (edges) |weight, vi| {
            if (dist[vi] > dist[ui] + weight) {
                return .{ .dist = dist, .found_negative = false };
            }
        }
    }

    return .{ .dist = dist, .found_negative = true };
}

pub fn floydWarshall(allocator: std.mem.Allocator, adj_matr: [][]f64) ![][]f64 {
    const n = adj_matr.len;

    var dist: [][]f64 = try allocator.alloc([]f64, n);
    for (dist) |*row, ui| {
        row.* = try allocator.alloc(f64, n);
        for (row.*) |*col, vi| {
            if (ui == vi) {
                col.* = 0;
            } else if (adj_matr[ui][vi] == 0) {
                col.* = std.math.inf(f64);
            } else {
                col.* = adj_matr[ui][vi];
            }
        }
    }

    var i: usize = 0;
    while (i < n) : (i += 1) {
        for (adj_matr) |edges, ui| {
            for (edges) |_, vi| {
                const new_dist = dist[ui][i] + dist[i][vi];
                if (dist[ui][vi] > new_dist) {
                    dist[ui][vi] = new_dist;
                }
            }
        }
    }

    return dist;
}

File: mat/1/src/path.cpp
#include "path.hpp"

#include <limits>
#include <queue>

class Vertex {
public:
	Vertex(size_t index, double dist)
	{
		this->m_index = index;
		this->m_dist = dist;
	}

	size_t index() const
	{
		return m_index;
	}

	double dist() const
	{
		return m_dist;
	}

	bool operator>(const Vertex &rhs)
	{
		return m_dist > rhs.m_dist;
	}

private:
	size_t m_index;
	double m_dist;
};

std::vector<double> dijkstra(const SquareMatrix<double> &adj_matr, size_t start)
{
	size_t n = adj_matr.size();

	std::vector<bool> visited(n);
	std::fill(visited.begin(), visited.end(), false);

	std::vector<double> dist(n);
	std::fill(dist.begin(), dist.end(), std::numeric_limits<double>::infinity());
	dist[start] = 0;

	std::priority_queue<Vertex, std::vector<Vertex>, std::greater<>> pqueue;
	pqueue.push(Vertex(start, 0));

	while (!pqueue.empty()) {
		Vertex current = std::move(pqueue.top());
		pqueue.pop();

		size_t i = current.index();

		if (visited[i]) {
			continue;
		}

		visited[i] = true;

		for (size_t j = 0; j < n; ++j) {
			if (adj_matr[i, j] == 0) {
				continue;
			}

			double new_dist = dist[i] + adj_matr[i, j];
			if (new_dist < dist[j]) {
				dist[j] = new_dist;
				pqueue.push(Vertex(j, new_dist));
			}
		}
	}

	return dist;
}

std::pair<std::vector<double>, bool> ford_bellman(const SquareMatrix<double> &adj_matr, size_t start)
{
	size_t n = adj_matr.size();

	std::vector<double> dist(n);
	std::fill(dist.begin(), dist.end(), std::numeric_limits<double>::infinity());
	dist[start] = 0;

	for (size_t k = 0; k < n - 1; ++k) {
		for (size_t i = 0; i < n; ++i) {
			for (size_t j = 0; j < n; ++j) {
				if (adj_matr[i, j] == 0) {
					continue;
				}

				double new_dist = dist[i] + adj_matr[i, j];
				if (new_dist < dist[j]) {
					dist[j] = new_dist;
				}
			}
		}
	}

	for (size_t i = 0; i < n; ++i) {
		for (size_t j = 0; j < n; ++j) {
			if (dist[i] > dist[j] + adj_matr[i, j]) {
				return { dist, false };
			}
		}
	}

	return { dist, true };
}

SquareMatrix<double> floyd_warshall(const SquareMatrix<double> &adj_matr)
{
	size_t n = adj_matr.size();
	SquareMatrix<double> dist(n);

	for (size_t i = 0; i < n; ++i) {
		for (size_t j = 0; j < n; ++j) {
			if (i == j) {
				dist[i, j] = 0;
			} else if (adj_matr[i, j] == 0) {
				dist[i, j] = std::numeric_limits<double>::infinity();
			} else {
				dist[i, j] = adj_matr[i, j];
			}
		}
	}

	for (size_t k = 0; k < n; ++k) {
		for (size_t i = 0; i < n; ++i) {
			for (size_t j = 0; j < n; ++j) {
				double new_dist = dist[i, k] + dist[k, j];
				if (dist[i, j] > new_dist) {
					dist[i, j] = new_dist;
				}
			}
		}
	}

	return dist;
}
